vertx-web/src/main/java/io/vertx/ext/web/handler/sse/EventSourceImpl.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2020 Red Hat, Inc.
+ *
+ * Red Hat licenses this file to you under the Apache License, version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.vertx.ext.web.handler.sse;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.VertxException;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpClient;
+import io.vertx.core.http.HttpClientRequest;
+import io.vertx.core.http.HttpClientResponse;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.http.HttpMethod;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class EventSourceImpl implements EventSource {
+
+  private HttpClient client;
+  private boolean connected;
+  private String lastId;
+  private Handler<String> messageHandler;
+  private final Map<String, Handler<String>> eventHandlers;
+  private SSEPacket currentPacket;
+  private final Vertx vertx;
+  private final EventSourceOptions options;
+  private Long retryTimerId;
+
+  public EventSourceImpl(Vertx vertx, EventSourceOptions options) {
+    options.setKeepAlive(true);
+    this.vertx = vertx;
+    this.options = options;
+    eventHandlers = new HashMap<>();
+  }
+
+  @Override
+  public synchronized EventSource connect(String path, Handler<AsyncResult<Void>> handler) {
+    return connect(path, null, handler);
+  }
+
+  @Override
+  public synchronized EventSource connect(String path, String lastEventId, Handler<AsyncResult<Void>> handler) {
+    if (connected) {
+      throw new VertxException("SSEConnection already connected");
+    }
+    if (client == null) {
+      client = vertx.createHttpClient(options);
+    }
+    client.redirectHandler(resp -> {
+      String redirect = resp.headers().get(HttpHeaders.LOCATION);
+      return Future.succeededFuture(createRequest(redirect, lastEventId, handler));
+    });
+    createRequest(path, lastEventId, handler).end();
+    return this;
+  }
+
+  @Override
+  public synchronized void close() {
+    if (retryTimerId != null) {
+      vertx.cancelTimer(retryTimerId);
+      retryTimerId = null;
+    }
+    if (client != null) {
```

Comment:
move the client close outside of the synchronized block (after copying the reference to a local variable and setting the ref to null)

Reply:
To clarify, do you mean something like this?

```java
  @Override
  public void close() {
    HttpClient client;
    synchronized (this) {
      if (retryTimerId != null) {
        vertx.cancelTimer(retryTimerId);
        retryTimerId = null;
      }

      client = this.client;
      this.client = null;
      connected = false;
    }

    if (client != null) {
      try {
        client.close();
      } catch (Exception e) {
        log.error("An error occurred closing the EventSource: ", e);
      }
    }
  }
```